net/connstats/stats.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package connstats maintains statistics about connections
+// flowing through a TUN device (which operate at the IP layer).
+package connstats
+
+import (
+	"net/netip"
+	"sync"
+
+	"tailscale.com/net/packet"
+	"tailscale.com/types/netlogtype"
+)
+
+// Statistics maintains counters for every connection.
+// All methods are safe for concurrent use.
+// The zero value is ready for use.
+type Statistics struct {
+	mu       sync.Mutex
+	virtual  map[netlogtype.Connection]netlogtype.Counts
+	physical map[netlogtype.Connection]netlogtype.Counts
+}
+
+// UpdateTxVirtual updates the counters for a transmitted IP packet
+// The source and destination of the packet directly correspond with
+// the source and destination in netlogtype.Connection.
+func (s *Statistics) UpdateTxVirtual(b []byte) {
+	s.updateVirtual(b, false)
+}
+
+// UpdateRxVirtual updates the counters for a received IP packet.
+// The source and destination of the packet are inverted with respect to
+// the source and destination in netlogtype.Connection.
+func (s *Statistics) UpdateRxVirtual(b []byte) {
+	s.updateVirtual(b, true)
+}
+
+func (s *Statistics) updateVirtual(b []byte, receive bool) {
```

Comment:
GitHub is being silly, this is identical to the removed `tunstats.Statististics.update` method.